rfcs/0000-support-injected-sidecars.md

```diff
@@ -0,0 +1,124 @@
+builds are unable to reach the network because they are attempting to communicate through the envoy proxy container that
+has not and will not start until the build completes.
+
+To get around this issue, kpack currently applies an annotation (`sidecar.istio.io/inject=false`) to all build pods to
+disable Istio sidecar injection. This presents a problem where users might want to take advantage of the various
+features of Istio, but cannot because kpack is bypassing it.
+
+[buildpack]: https://buildpacks.io/
+
+[cnb-lifecycle]: https://buildpacks.io/docs/concepts/components/lifecycle/
+
+[envoy-proxy]: https://www.envoyproxy.io/
+
+[init-containers]: https://kubernetes.io/docs/concepts/workloads/pods/init-containers/
+
+[istio]: https://istio.io/
+
+[istio-sidecar-injection]: https://istio.io/latest/docs/setup/additional-setup/sidecar-injection/#injection
+
+[istio-service-mesh]: https://istio.io/latest/about/service-mesh/#what-is-a-service-mesh
+
+[istio-init-container-compatability]: https://istio.io/latest/docs/setup/additional-setup/cni/#compatibility-with-application-init-containers
+
+## Outcome
+
+When kpack is running with `SUPPORT_INJECTED_SIDECARS=true`, builds can execute successfully in environments where
+sidecars are executed into the build pods (i.e. istio proxy sidecar injection).
+
+## Actions to take
+
+1. Create a new binary that can be used as a new entrypoint for build steps. The purpose of this binary will be to allow
+   for specific build steps to depend on other build steps to complete before executing since the containers will be
+   running concurrently. There will also need to be a mechanism for propagating errors in one build step so that a build
+   can exit on error.
+
+   Example of possible implementation is in the [spike][kpack-spike-build-waiter].
+2. Convert the existing init containers in to be regular containers and add modify the entrypoint to be the new binary.
+3. We will put this new behavior behind a feature gate (Environment variable added to kpack controller deployment
+   called `SUPPORT_INJECTED_SIDECARS`) and only execute in regular containers if enabled by the user. In
+   the event that the feature flag is not set or set to false, kpack will continue to execute builds in init containers.
+4. Due to the fact that sidecars such as istio-proxy, do not exit when all other containers exit, we will need to add a
+   method for stopping the proxy once a build exits
+    1. To do this we can update the running pod's `spec.activeDeadlineSeconds` to 1 which will cause the pod to go into
+       a failed state and kill all containers after 1 second.
+    2. This will kill the sidecar, but it will also leave our pod in a failed state with the message `DeadlineExceeded`,
+       leading the reconciler to believe the build has failed. To circumvent this we will need allow pods to
+       fail with the reason `DeadlineExceeded` as long the completion container has successfully completed.
```

Comment:
Is there a different way to solve this?
I'm not sure it's a good idea to allow pods to fail with the reason `DeadlineExceeded`, since we may find in the future, a different flow that shouldn't be ignored with the same error.

Reply:
The alternative is to replace the image the sidecar is using with a no-op image that just exits (this is what tekton does). The problem with that is that the istio sidecar adds a lifecycle prehook (which we can't remove) that attempts to execute a particular command that isn't present in the no-op image so it still ends up exiting with an error. Another alternative would be to sent a POST request to the istio sidecar to tell it to shutdown, but I decided not to go down this path because it is very istio specific whereas the current proposal would work for any sidecar. I'm not super concerned about the DeadlineExceeded error because we are checking that the build still completed successfully before marking the build as successful. If the build failed and causes a deadline exceeded it will still fail the build.